lib/render/RayCaster.cpp

```diff
@@ -430,41 +480,88 @@ bool RayCaster::UserCoordinates::UpdateCoordinates( const RayCasterParams* param
     {
         for( size_t i = 0; i < numOfVertices; i++ )
         {
-            dataField[ i ] = (float(*valItr) - valueRange[0]) * valueRange1o;
+            dataField[ i ] = float(*valItr);
             ++valItr;
         }
     }
 
     delete grid;
-    return true;
+    return 0;
 }
 
-int RayCaster::_initializeGL()
+int  RayCaster::UserCoordinates::UpdateCurviCoords( const RayCasterParams* params,
+                                                          DataMgr*         dataMgr )
 {
-#ifdef Darwin
-    return 0;
-#endif
-    // Enable debug output
-    glEnable              ( GL_DEBUG_OUTPUT );
-    glDebugMessageCallback( MessageCallback, 0 );
-
-    _loadShaders();
+    assert( params->GetCastingMode() == 2 );
+
+    if( xyCoords )
+        delete[] xyCoords;
+    xyCoords = new float[ dims[0] * dims[1] * 2 ];
+    if( zCoords )
+        delete[] zCoords;
+    zCoords  = new float[ dims[0] * dims[1] * dims[2] ];
+    if( !zCoords )  // Test if allocation successful for 3D buffers. 
+    {
+        MyBase::SetErrMsg("Failed to allocate memory");
+        return -1;
+    }
 
-    // Create Vertex Array Object (VAO)
-    glGenVertexArrays( 1, &_vertexArrayId );
-    glGenBuffers(      1, &_vertexBufferId );
-    glGenBuffers(      1, &_indexBufferId );
+    // Gather the XY coordinate from frontFace buffer
+    size_t xyIdx = 0, xyzIdx = 0;
+    for( size_t y = 0; y < dims[1]; y++ )
+        for( size_t x = 0; x < dims[0]; x++ )
+        {
+            xyCoords[ xyIdx++ ] = frontFace[xyzIdx++];
+            xyCoords[ xyIdx++ ] = frontFace[xyzIdx++];
+            xyzIdx++;
+        }
 
-    _initializeFramebufferTextures();
+    // Gather the Z coordinates from grid
+    StructuredGrid* grid = nullptr;
+    if( this->GetCurrentGrid( params, dataMgr, &grid ) != 0 )
+    {
+        MyBase::SetErrMsg("Failed to retrieve a StructuredGrid");
+        return 1;
+    }
+    StructuredGrid::ConstCoordItr coordItr   = grid->ConstCoordBegin();
+    size_t numOfVertices = dims[0] * dims[1] * dims[2];
+    for( xyzIdx = 0; xyzIdx < numOfVertices; xyzIdx++ )
+    {
+        zCoords[xyzIdx] = float((*coordItr)[2]);
+        ++coordItr;
+    }
 
     return 0;
 }
 
+int RayCaster::_initializeGL()
+{
+    if( _loadShaders() != 0 )
+    {
+        MyBase::SetErrMsg("Failed to load shaders!");
+        return -1;
+    }
+    if( _initializeFramebufferTextures() != 0 )
+    {
+        MyBase::SetErrMsg("Failed to Create Framebuffer and Textures!");
+        return -1;
+    }
+
+    return 0;   // Success
+}
+
 int RayCaster::_paintGL( bool fast ) 
 {
-#ifdef Darwin
-    return 0;
+#ifdef DEBUG
```

Comment:
I think you mean NDEBUG. DEBUG is not a standard macro. Also, this is automatically disabled on release builds.

Reply:
I intended to disable that code on release build, and only enable it on debug build. Does it make sense?

Reply:
Yes. The macro to do that is NDEBUG. Also, the ShaderManager class already implements that functionality.

Reply:
This is addressed in 41b11793.

Reply:
Needs to follow shader initialization:

`    SmartShaderProgram shader = _glManager->shaderManager->GetShader("DVRsomething");
    if (!shader.IsValid())
        return -1;`

...
followed by shader unbind at any return point from paintGL

Reply:
@stasj will address